lib/hubspot/resource.rb

```diff
@@ -0,0 +1,183 @@
+module Hubspot
+  class Resource
+
+    class_attribute :id_field, instance_writer: false, default: "id"
+    class_attribute :property_name_field, instance_writer: false, default: "property"
+    class_attribute :update_method, instance_writer: false, default: "put"
+
+    class << self
+      def find(id)
+        instance = new(id)
+        instance.reload
+      end
+
+      def create(properties = {})
+        request = {
+          properties: Hubspot::Utils.hash_to_properties(properties.stringify_keys, key_name: property_name_field)
+        }
+        response = Hubspot::Connection.post_json(create_path, params: {}, body: request)
+        new(response[id_field], response)
+      end
+    end
+
+    def initialize(id = nil, response = {})
+      response = response.with_indifferent_access
+
+      @id = id
+      initialize_from(response)
+
+      @persisted = @id.present?
+      @deleted = false
+      @changes = {}
+    end
+
+    def id
+      @id
+    end
+
+    def id=(id)
+      @id = id
+    end
+
+    def metadata
+      @metadata
+    end
+
+    def changes
+      @changes
+    end
+
+    def [](name)
+      @properties[name]
+    end
+
+    def reload
+      response = Hubspot::Connection.get_json(find_path, id: @id)
+      initialize_from(response.with_indifferent_access)
+
+      self
+    end
+
+    def persisted?
+      @persisted
+    end
+
+    def save
```

Comment:
What do you think of simplifying `save` by exposing an `update(attributes)` function? I like the obviousness that accompanies the exposure of an `update` method and it would allow users to pass in multiple parameters.

This would change the user API from:
```
contact.firstname = "FirstName"
contact.lastname = "LastName"
contact.save
```
to
```
contact.update(firstname: "FirstName", lastname: "LastName")
```

For the `create` action, I'm curious if you think users will want to construct a `Hubspot::Contact` object and then call the instance method `save` or if it seems reasonable to define `save|create` as a class method only, which returns an instance of `Hubspot::Create`.

Reply:
I think like [ActiveRecord](https://guides.rubyonrails.org/active_record_basics.html) we should have both a `save` and `update` method for the instance. This way users can set several attributes and have the update performed with one call or can set properties and later call `save`. This way the API is flexible for whichever form fits their code the best.

Just like ActiveRecord the `create` method would be used to specify one or more properties and would call the API to create the object and return the resource. If instead a user wanted to construct a contact, they can call `new`, set the properties, and when finished call `save`. `save` can't be a class method because there is no state to save.